apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -790,6 +774,51 @@ wait_for_session_or_sm({xmlstreamerror, _}, StateData) ->
 wait_for_session_or_sm(closed, StateData) ->
     {stop, normal, StateData}.
 
+maybe_do_compress(El, NextState, StateData) ->
+    SockMod = (StateData#state.sockmod):get_sockmod(StateData#state.socket),
+    {Zlib, _} = StateData#state.zlib,
+    #xmlel{name = Name, attrs = Attrs, children = _} = El,
+    case {xml:get_attr_s(<<"xmlns">>, Attrs), Name} of
+        {?NS_COMPRESS_BIN, <<"compress">>} when Zlib == true,
+                                                ((SockMod == gen_tcp) or
+                                                 (SockMod == ejabberd_tls)) ->
+          check_compression_auth(El, NextState, StateData);
+        _ ->
+            process_unauthenticated_stanza(StateData, El),
+            fsm_next_state(NextState, StateData)
+
+    end.
+
+check_compression_auth(_El, NextState, StateData) ->
+  Auth = StateData#state.authenticated,
+  case Auth of
+    false ->
+      send_element(StateData, compress_setup_failed()),
+      fsm_next_state(NextState, StateData);
+    _ ->
+      check_compression_method(_El, NextState, StateData)
+  end.
+
+check_compression_method(El, NextState, StateData) ->
+  case xml:get_subtag(El, <<"method">>) of
+    false ->
+      send_element(StateData, compress_setup_failed()),
+      fsm_next_state(NextState, StateData);
+    Method ->
+      case xml:get_tag_cdata(Method) of
```

Comment:
I think we can use `exml_query:path/2` function here to look for the method in line `803`. Doing that, we'll have only one case clause.